### PR TITLE
[WIP] purify: new package

### PR DESCRIPTION
**Pull request by tkoskela**
Add package for astrophysics imaging application [purify](https://github.com/astro-informatics/purify/releases/tag/v4.2.0). Depends on #46837 and #46838

### CODE REVIEW

**Comment by tkoskela** [comment]
@spackbot commands

**Comment by tkoskela** [comment]
@spackbot re-run pipeline

**Comment by tkoskela** [comment]
Ok, it looks like my merge commit re-ran the CI anyways :see_no_evil: It looks like it now finds all the dependencies :tada:

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
This is only a preview of my results:
```py
bin/spack find -v purify | sed 's/~[a-z]*//g;s/ [a-z0-9_]*=[a-zA-Z0-9]*//g'
-- linux-ubuntu22.04-skylake / gcc@13.1.0 -----------------------
purify@4.2.0+tests
purify@4.2.0+openmp
purify@4.2.0+openmp+tests
purify@4.2.0+mpi
purify@4.2.0+docs
purify@4.2.0+coverage
purify@4.2.0+benchmarks
```
But these builds were only possible after some fixes and updates.
I'll report on those in separate comments.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
When my build of `sopt` picked the latest `libtiff@4.6.0`, it broke the build:
```py
-- Found TIFF: .../libtiff-4.6.0-.../lib/cmake/tiff (found version "4.6.0")
-- Configuring done (1.6s)
CMake Error in cpp/purify/CMakeLists.txt:
  Imported target "sopt" includes non-existent path

    ".../libtiff-4.6.0-.../lib/cmake/tiff/include"
  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:
  * The path was deleted, renamed, or moved to another location.
  * An install or uninstall procedure did not complete successfully.
  * The installation package was faulty and references files it does not provide.
```
Option 3 is the cause, see these lines from the log:
```py
-- Found TIFF: .../libtiff-4.6.0-.../lib/cmake/tiff ...
    ".../libtiff-4.6.0-.../lib/cmake/tiff/include"
```
The 2nd line is indeed an incorrect include path.
The correct include path is `.../libtiff-4.6.0-.../include`
It must be (without `lib/cmake/tiff`) as this path only has the `cmake` Modules for `tiff`.

The cause is that `tiff` upstream messed this up:
http://www.simplesystems.org/libtiff/releases/v4.6.0.html
- TiffConfig.cmake.in: set TIFF_INCLUDE_DIR, TIFF_INCLUDE_DIRS and TIFF_LIBRARIES for compatibility with FindTIFF. cmake as shipped by CMake (fixes [issue #589](https://gitlab.com/libtiff/libtiff/-/issues/589))

Fixed in 4.7.0: http://www.simplesystems.org/libtiff/releases/v4.7.0.html
- CMake: Fix TIFF_INCLUDE_DIRS

So tiff-4.6.0 does not work in this case:
- sopt gets the wrong include path by importing the INTERFACE_INCLUDE_DIRECTORIES from tiff and re-exports it ins own include directories for its users, where purify finally fails with this error as the directory is wrong.

So the fix is to update to 4.7.0 and exclude 4.6.0 from being used by `sopt`. It's an easy fix. New version and update the `depends_on` in sopt (or add a conflict for 4.6.0 in case 4.5.0 works, and you absolutely need 4.5.0 and 4.7.0 is fine too - but simply raising the range to "libtiff@4.7:" is easier). I've a tested diff/commit for that.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
There is another issue. Which is why I blocked GitLab CI to prevent compute costs by setting the PR do draft state first.

- `sopt/purity+onnxrt` needs `onnxrtuntime`, which spack provides as `py-onnxruntime`.

If `sopt+onnxrt` is not requested when `purity+onnxrt` is build, you run into this error that tells you: `message(FATAL_ERROR "SOPT built without ONNXrt support")`:


https://github.com/astro-informatics/purify/blob/development/cmake_files/dependencies.cmake#L39

The `README.md` also tells that [ONNXruntime](https://onnxruntime.ai/) v1.17.1 is required for onnxrt support:
> https://github.com/astro-informatics/purify?tab=readme-ov-file#dependencies-installation
> https://github.com/astro-informatics/sopt?tab=readme-ov-file#dependencies-installation
> - [ONNXruntime](https://onnxruntime.ai/) v1.17.1 - Optional - a cross-platform runtime engine based on the Open Neural Network eXchange format. sets. The main PURIFY program requires this library (and its dependencies)

While `sopt` downloads and extracts an older, precompiled `libonnxruntime` from the Microsoft server, as fallback in case it is not provided with an installed onnxruntime, this fallback is far from ideal, and it is not found by `purtiy` either.

Instead, use the `onnxrtruntime` built from source, optimized for the target platform as `py-onnxruntime`.

For `sopt`, simply adding the needed `depends_on("py-onnxruntime@1.17:1", when="+onnxrt")` appears to be enough that it's cmake file detects it and does not download and extract the external, precompiled `libonnxruntime` from the Microsoft server anymore.

**Review by bernhardkaindl — COMMENTED** [review]
So, with these changes (adding the needed onnxruntime package, also defining PURITY_ONNXRT for cpp), purity's onnxrt code is attempted to be compiled, but at least with the current versions and gcc-13.1, the build fails as described in https://github.com/astro-informatics/purify/pull/360.

Unless a solution for this is found, it may be better to disable/comment-out the `+onnxrt` variant in `purity/package.py` and enable it later when the issue is fixed.

**Comment by tkoskela** [comment]
Thank you for your review and suggestions @bernhardkaindl! Sorry about not thoroughly testing the onnx variant. We're currently working on updates on it in purify, so I think your suggestion to remove the variant from this PR is a good one. I'll add it back when we've resolved the build issues.

**Review by bernhardkaindl — DISMISSED** [review]

File: var/spack/repos/builtin/packages/libtiff/package.py [hunk 1]
```diff
@@ -40,6 +40,7 @@ class Libtiff(CMakePackage, AutotoolsPackage):
 
     license("libtiff")
 
+    version("4.7.0", sha256="67160e3457365ab96c5b3286a0903aa6e78bdc44c4bc737d2e486bcecb6ba976")
```
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/libtiff/package.py]: #46999 adds 4.7.0 as well:
```suggestion
```
The GitLab CI pipelines of #46999 is currently running.

After it is merged, we can commit the suggestion as 4.7.0 is then existing upstream,
and rebase the PR to the merged commit of #46999 on develop.

I'd then approve and enable auto-merge.

If all goes well, is would be possible to merge it this week.

**Review by bernhardkaindl — APPROVED** [review]

**Pull request merged**